<?xml version="1.0"?>

<launch>

    <arg name="use_sim" default="false" />

    <arg name="launch_camera" default="true" />

    <arg name="launch_cam_tf_publisher" default="true" />

    <arg name="robot_id" default="both" />

    <arg name="gripper" default="rg2_ft" />

    
    <arg name="host_ip"  default="10.0.0.3"/> <!-- 10.0.0.3 or 127.0.0.1 -->
    <arg name="robot_ip"  default="10.0.0.1"/> <!-- 10.0.0.1 or 127.0.0.1 -->
    <arg name="reverse_port"  default="8080"/> <!-- 50001 -->

    <arg name="move2jpos_topic"  default="move2jpos_service"/>
    <arg name="tf_publish_topic"  default="/robot_camera_tf"/>

    <param name="robot_description" command="$(find xacro)/xacro '$(find im_traj_demo_controller)/urdf/record_demo.xacro' robot_id:=$(arg robot_id) gripper:=$(arg gripper) load_camera:=true" />
    <node name="rviz" pkg="rviz" type="rviz" respawn="false" args="-d $(find im_traj_demo_controller)/rviz/record_demo.rviz" output="screen" required="true"/>

    <!-- <node name="robot_joint_state_publisher" pkg="joint_state_publisher_gui" type="joint_state_publisher_gui">
      <remap from="robot_description" to="robot_description" />
    </node>

    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher">
      <remap from="robot_description" to="robot_description" />
    </node>
     -->

    <arg name="camera" default="rs2" />
    <group if="$(arg launch_camera)">
      <include file="$(find realsense2_camera)/launch/rs_camera.launch">
        <arg name="camera" value="$(arg camera)" /> <!-- This will be the prefix of the published image and tags header -->

        <arg name="publish_tf" value="false"/>

        <arg name="enable_depth" value="true" />
        <arg name="align_depth" value="true" />
        <arg name="enable_sync" value="true" />
        <arg name="enable_pointcloud" value="true" />
        
        <arg name="color_width" value="640"/>
        <arg name="color_height" value="480"/>
        <arg name="color_fps" value="30"/>

        <arg name="depth_width" value="640"/>
        <arg name="depth_height" value="480"/>
        <arg name="depth_fps" value="30"/>
    </include>

    </group>

    <!-- launch-prefix="xterm -fa monaco -fs 13 -e gdb -args" -->
    <arg name="node_name" default="project_name__node" />
    <node pkg="main_program" type="main_program" name="$(arg node_name)"
          output="screen" required="true">

        <param name="use_sim" value="$(arg use_sim)"/>

        <param name="host_ip" type="string" value="$(arg host_ip)" />
        <param name="robot_ip" type="string" value="$(arg robot_ip)" />
        <param name="reverse_port" type="int" value="$(arg reverse_port)" />

        <param name="move2jpos_topic" type="string" value="$(arg move2jpos_topic)" />
        <param name="robot_cam_tf_topic" type="string" value="$(arg tf_publish_topic)" />

        <param name="image_topic" type="string" value="/$(arg camera)/color/image_raw" />
        <!-- <param name="depth_image_topic" type="string" value="/$(arg camera)/depth/image_rect_raw" /> -->
        <param name="depth_image_topic" type="string" value="/$(arg camera)/aligned_depth_to_color/image_raw" />
        <param name="cam_info_topic" type="string" value="/$(arg camera)/color/camera_info" />

        <rosparam>

            im_traj_demo_controller: true

            robot_type: "ur"
            robot_description_name: "/robot_description"
            base_link: "robotic_platform_base_link"
            ftsensor_link: "ur5_right_tool0"
            tool_link: "ee_link"
            Fext_dead_zone: [3.0, 3.0, 3.0,  0.3, 0.3, 0.3]

            ctrl_cycle: 0.002
            limits_check: false

            q_start: [-0.41, -1.51, -1.82, -3.57, 3.93, -1.14]

            r85_gripper_options: 
              {use_sim: false, prefix: "rg2_ft", http_link: "http://192.168.1.1:80"}

        </rosparam>

    </node>

    <node pkg="robot_camera_tf_publisher" type="robot_camera_tf_publisher" name="robot_camera_tf_publisher_node"
          output="screen" cwd="node" required="true" if="$(arg launch_cam_tf_publisher)">

        <param name="use_sim" value="$(arg use_sim)"/>

        <param name="move2jpos_topic" type="string" value="$(arg move2jpos_topic)" />
        <param name="tf_publish_topic" type="string" value="$(arg tf_publish_topic)" />

        <rosparam>

            host_ip: "10.0.1.3"
            robot_ip: "10.0.1.1"
            reverse_port: 8081
        
            robot_type: "ur"
            robot_description_name: "/robot_description"
            base_link: "robotic_platform_base_link"
            camera_color_optical_link: "rs2_color_optical_frame"

            # q_start: [-4.13, -2.4, -0.912, -1.85, 2.37, -0.96]
            
        </rosparam>

    </node>

</launch>
